File: Graph(그래프)/백준_10451_순열사이클/순열 사이클.py
arr = [0 for _ in range(1001)]
check = [0 for _ in range(1001)]

def cycle(n):
    count = 0
    for i in range(1, n+1):
        if check[i] == 0:
            count+=1
            start = i
            end = arr[i]
            while start != end:
                check[end] = 1
                end = arr[end]
    return count

T = int(input())
input_arr = []
while 1:
    T-=1
    if T == -1: break
    n = int(input())
    input_arr = list(map(int, input().split()))
    for i in range(1,n+1):
        arr[i] = input_arr[i-1]
        check[i] = 0
    print(cycle(n))

File: Graph(그래프)/백준_10451_순열사이클/순열 사이클.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int arr[1001];
bool check[1001];

int cycle(int n) {
	int count = 0;
	for (int i = 1; i <= n; i++) {
		if (check[i] == false) {
			count++;
			int start = i;
			int end = arr[i];
			while (start != end) {
				check[end] = true;
				end = arr[end];
			}
		}
	}
	return count;
}

int main() {
	int T = 0;
	cin >> T;
	int n = 0;
	while (T--) {
		cin >> n;
		for (int i = 1; i <= n; i++) {
			cin >> arr[i];
			check[i] = false;
		}
		cout << cycle(n) << "\n";
	}

}